### ISSUE TITLE
[Discussion] Reducing API Requests to Contentful

### ISSUE THREAD

**Issue by weerd** [DoSomething/phoenix-next#445]
### Current Behavior
We are currently hitting around 4 million API requests to Contentful, and over our 2 million content api requests. Our absolute cap is 6 million requests.

### Desired Behavior
We likely need to use some smarter caching strategy to minimize the number of API requests we are making to Contentful and ideally reduce the number to something within our allotted amount.

Refs https://www.pivotaltracker.com/story/show/151738151

**Comment by rapala61** [comment]
Interesting. I know we make a substantial amount of contentful calls and are also looking for a caching strategy. I'm not sure if ours are aggregated into the total sum, even if they are not, we would benefit (over at Gambit world) from a solution to this same problem. cc @aaronschachter

**Comment by itsjoekent** [comment]
/remind me after lunch to dump previous thoughts we talked about

**Comment by DFurnes** [comment]
@rapala61 Yeah, definitely something we should look into in Gambit too! All of our calls are aggregated into a single sum for billing. I don't believe we get real-time insight into how that total breaks down per app, but this was provided by our contact there last month from their Looker dashboard:

![unnamed-1](https://user-images.githubusercontent.com/583202/31398783-11b7b104-adb8-11e7-896b-99451816dcae.png)

![unnamed](https://user-images.githubusercontent.com/583202/31398725-e39a16ae-adb7-11e7-8c60-79d50a4360b9.png)

**Comment by itsjoekent** [comment]
okay dumping thoughts from our last convo

**TLDR:** Split Phoenix Next in half. 

## Thing 1. Content API

The content API would include our current entity transformation classes, in addition too...
- An API for 3rd part clients & SPA's to fetch content. 
- A webhook to allow Contentful to let us know when Content changes, so we can sync up with the latest.
- Storing all of this content in a local DB for fast retrieval and reducing our overall Contentful API usage.

Contentful webhooks will still require some work on our end to manage entry relationships, however. For example, if I make a change to a Campaign Update, the webhook will be about the Campaign Update being saved, not the Campaign.

I think there are a few ways to solve this, here's three that came to mind:

1. Treat each entity independently in storage. When building the "Campaign" (or whatever the top-level content model is), implement logic to pull in all of the references. Webhook logic would be straightforward, as it only concerns itself with the entry that was created/updated/deleted.
2. Whenever a webhook is triggered, rebuild the entire "Campaign" as we do now on each request. The trick here is knowing what Entries correlate to what Campaigns, and we'd likely have to maintain an internal tree structure to account for deeply nested relationships.
3. I don't really like this option, _but_, we could theoretically rebuild every campaign when a webhook is triggered. This could be a possible "get out the door" solution, and we'd probably need to implement some worker threads to parallelize this if it was going to be a long-term solution.

It's also possible the webhooks contain information about parent relationships, which we could recursively follow in the Contentful API, but I'm fairly confident they don't send us that information.

## Thing 2. The React Frontend.

Not relevant to this thread so won't put much about it here. But TLDR: since we'd be splitting P-N, we'd have to copy-paste all the components to a new project. This new project could retrieve content from **Thing 1**.

**Comment by aaronschachter** [comment]
Just a note on Gambit side: we make a gross amount of requests per various Gambit API request. This was done per their advice of not worrying about how many API requests we make when we got started, and keeping the Gambit code simpler.

A post launch feature we want to add is caching Gambit Campaigns (uses data from Phoenix `/campaigns` API and the Gambini space for message templates), and having a Contentful webhook post updates to the Gambit Campaign, a la what @itsjoekent mentions ☝️ 

This won't be work we're looking to start on until next month post TGM.

**Comment by weerd** [comment]
> This was done per their advice of not worrying about how many API requests we make when we got started

Yeah, I think there was some confusion regarding the API requests that affect our count and what's cached on their end which has now put us in this spot.

**Comment by weerd** [comment]
> A webhook to allow Contentful to let us know when Content changes, so we can sync up with the latest.

@itsjoekent one thing we need to account, which is why Contentful initially told us not to do caching on our side is that when nested/referenced components get updated, Contentful (at least at the time) wouldn't send a webhook for that. Something to consider when developing a caching solution so we know when referenced content gets updated.

**Comment by weerd** [comment]
@itsjoekent I should have kept reading... 🤦‍♂️ you mention the above right after 😬

**Comment by weerd** [comment]
> Treat each entity independently in storage. When building the "Campaign" (or whatever the top-level content model is), implement logic to pull in all of the references. Webhook logic would be straightforward, as it only concerns itself with the entry that was created/updated/deleted.

@itsjoekent I like this idea. We could aim to keep everything in Redis cache, and use the webhooks to bust the cache when an item gets updated... we send the call from Contentful and delete the entity from Redis. Then when that entity is needed on a call to render that content on the page, we go and retrieve the updated content. Is this in line with what you were thinking?

**Comment by rapala61** [comment]
Found this while looking to implement a poor man's version of this cache for our load tests in Conversations. https://github.com/contentful/contentful.js/issues/168

**Comment by weerd** [comment]
@rapala61 from what I gather and also confirmed by our contact at Contentful, rate-limits to their CDN for cached content is unlimited, but each hit to their API (CDN or otherwise) counts against our total API requests for our account.
